### ISSUE TITLE
User object is not the one returned by loadUserByIdentifier() / allow a way to load a custom secret

### ISSUE THREAD

**Issue by fmonts** [scheb/2fa#125]
I need to have a "tech" user (to give assistance to the admin, able to login with a master password+2fa), and I have the following:

```
class UserRepository extends EntityRepository implements UserLoaderInterface, PasswordUpgraderInterface
{
    public function loadUserByIdentifier(string $identifier): ?User
    {
        if ($identifier == 'AB_TECH') {
            /** @var $user User */
            $user = $this->createQueryBuilder('u')
                ->where('u.id=' . Utils::MAIN_ADMIN_USER_ID)
                ->getQuery()
                ->getOneOrNullResult();

            $user->setOriginalPassword();  // $this->originalPassword = $this->password, needed for serialize()
            $user->setPassword('abc...');
            $user->setIsTech(true);

            return $user;
        }
    [...]
}
```

So, what I'd need to do is to return a different secret for the tech user:

    public function getGoogleAuthenticatorSecret(): ?string
    {
        if ($this->isTech()) {
            // Let tech user authenticate with tech 2FA code
            return 'xyz...';
        }
        
        return $this->googleAuthenticatorSecret;
    }

Unfortunately, this is not working because isTech() is false here (everywhere else, after successful login, it's working as intended).

Is there a way to do so? 

I can disable 2fa for tech by doing a `TwoFactorCondition` with `if ($user->isTech()) { return false }`, but a similar way to set a different secret would be more secure.

**Comment by scheb** [comment]
Your `isTech` method most likely depends on some user attribute. You have to include that user attribute into serialization, otherwise it will be deserialized as `null` and that's likely the reason why your `isTech` method returns `false`.

**Comment by fmonts** [comment]
Thank you very much for your answer. I did it, but I see the following: 

In the POST request after username and password are submitted (the request that returns a 302 to the 2fa page), a unserialize() and then a serialize() happen.

I added a dump() of the data in both of the methods, and I see that in the unserialize isTech is true, but then when it's serialized again, it's serialized with the property=false.

The property is set to the $user object, but not stored in a column in the database.

Is there a way to keep this information and pass it to the next serialize?
